Add branching priorities to CPLEXSHELL

ruaridhw opened the pull request:
## Fixes #1299 

## Summary/Motivation:
To be able to use the branching priority capability of commercial solvers when solving MIP problems

## Changes proposed in this PR:
- Include `branch_priority` and `branch_direction` attributes of variable data
- Read off these attributes as part of the CPLEXSHELL solve and write them to a CPLEX .ord file
- Update the CPLEX OPL script to read these .ord file should it exist

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

Commit:
:hammer: Add branching priorities to CPLEXSHELL

- Include `branch_priority` and `branch_direction` attributes of variable data
- Read off these attributes as part of the CPLEXSHELL solve and write them to a CPLEX .ord file
- Update the CPLEX OPL script to read these .ord file should it exist

Commit:
:rotating_light: Add tests of solving with priorities

- End-to-end tests of `opt.solve()` using branching priorities with CPLEXSHELL

Commit:
:hammer: Use `Suffix`es instead of `_GeneralVarData`

- Declare `priority` and `direction` suffixes on the model instead of attributes on the variable data in order to represent branching decisions

Commit:
:hammer: Allow indexed variables to have separate directions

Commit:
:bug: Handle both non-indexed vars in Suffix

mhusbyn commented on pyomo/solvers/plugins/solvers/CPLEX.py:
```diff
@@ -112,6 +113,36 @@ def __new__(cls, *args, **kwds):
+    default = 0
+    down = -1
+    up = 1
+
```
This could probably be pulled out to be a 'global' enum so it can be used for all solvers that support branching priorities. The only solver specific stuff here is the translation so that can be done in each individual solver.

That way you can build your model agnostic of what solver it will run on, which I don't think would be possible with the current implementation.

mhusbyn replied:
I see you have to specify a data type when declaring the Suffix. I guess you should add a 'data_type' attribute or something as well in that case that can be  used.

ruaridhw replied:
> The only solver specific stuff here is the translation

Ah not quite 😉 These magic integer values themselves are CPLEX-specific. See [here](https://www.ibm.com/support/knowledgecenter/SSSA5P_12.10.0/ilog.odms.cplex.help/CPLEX/Parameters/topics/BrDir.html). Having said that, it turns out that Gurobi also [just happens to share](https://www.gurobi.com/documentation/9.0/refman/branchdir.html) the same magic values... So I'll do the refactor!

mhusbyn replied:
Where are these magic integer values being used for CLEX specific stuff? Aren't they being translated in `to_str`?

ruaridhw replied:
As discussed, they will be used in the CPLEXDirect and GurobiDirect solver implementations which have not been included here. For now, I've removed `CPLEXBranchDirection` and moved the translation to `ORDFileSchema`

mhusbyn commented on pyomo/solvers/plugins/solvers/CPLEX.py:
```diff
@@ -112,6 +113,36 @@ def __new__(cls, *args, **kwds):
+                branch_direction
+            ]
+        except KeyError:
+            return ""
```
No idea if more efficient/readable but `dict.get` is an option here.  But Python tends to favour 'easier to ask for forgiveness' over 'look before you leap' so I guess this is fine too.

ruaridhw replied:
> But Python tends to favour 'easier to ask for forgiveness'

Yeah, that was my intention

mhusbyn commented on pyomo/solvers/plugins/solvers/CPLEX.py:
```diff
@@ -201,6 +232,66 @@ def _warm_start(self, instance):
+
+    def _write_priorities_file(self, instance) -> None:
+        """ Write a variable priorities file in the CPLEX ORD format. """
+        from pyomo.core.base import Var
```
Is there a reason this is an inline import?

ruaridhw replied:
I was copying `_warm_start()` where this is a local import too... I'll try removing both and 🤞 nothing breaks

mhusbyn commented on pyomo/solvers/plugins/solvers/CPLEX.py:
```diff
@@ -201,6 +232,66 @@ def _warm_start(self, instance):
+    SUFFIX_PRIORITY_NAME = "priority"
+    SUFFIX_DIRECTION_NAME = "direction"
+
+    def _write_priorities_file(self, instance) -> None:
```
NB: This has a return type (I'm assuming not Python 2.7 compliant)

ruaridhw replied:
Ah perfect, that's probably the reason 2.7 is breaking. Thanks!

mhusbyn commented on pyomo/solvers/plugins/solvers/CPLEX.py:
```diff
@@ -201,6 +232,66 @@ def _warm_start(self, instance):
+    SUFFIX_PRIORITY_NAME = "priority"
+    SUFFIX_DIRECTION_NAME = "direction"
+
+    def _write_priorities_file(self, instance) -> None:
```
What is instance? Is it worth documenting or is it Pyomo lingo?

ruaridhw replied:
Yup, it's standard (see `_warm_start()`)

mhusbyn commented on pyomo/solvers/plugins/solvers/CPLEX.py:
```diff
@@ -201,6 +232,66 @@ def _warm_start(self, instance):
+                        ORDFileSchema.ROW(byObject[id(child_var)], priority, child_var_direction)
+                    )
+
+            ord_file.write(ORDFileSchema.FOOTER)
```
Worth splitting this method up a bit? To me it seems like there's 2-3 steps here:
1. input preparation
2. translation
3. writing to the file

It looks to me like some of this logic is not CPLEX specific so if we manage to split this method up into parts it'll most likely be easier to implement extra solvers later. 

This is based on my assumption that it is mainly the _translation_ that is solver specific.

Commit:
:rotating_light: Update tests to work when CPLEX is unavailable

Commit:
:bug: Fix test

Commit:
:bug: Fix imports

- Looks like there's some circular imports failing specifically for Py2.7
- Follow the same (hacky) local import design chosen in 0935e75e

mhusbyn commented on pyomo/solvers/tests/checks/test_cplex.py:
```diff
@@ -54,5 +61,195 @@ def test_validate_file_name(self):
+            ' UP x11 2\n'
+            'ENDATA\n'
+        )
+
```
For these tests, please use the priority constant instead of magic numbers. That'll make them easier to maintain, e.g. if we decide to represent them with different numbers or something.

And there is no ambiguity of what they mean.

ruaridhw replied:
Updated, however I felt this was a bit too convoluted for `test_use_variable_priorities()`

mhusbyn commented on pyomo/solvers/tests/checks/test_cplex.py:
```diff
@@ -54,5 +61,195 @@ def test_validate_file_name(self):
+            ' DN x10 2\n'
+            ' UP x11 2\n'
+            'ENDATA\n'
+        )
```
Perhaps this assertion is neater if you read the contents of `priorities_file_name` before doing the final solve and comparing that against `priorities_file`? They should be the same right?

ruaridhw replied:
Sure

mhusbyn reviewed (changes requested):
See comments. Main bits:
- Pulling out direction class into a more global enum and making the translation CPLEX specific
- Splitting up `_write_priorities_file` (if it makes sense)

Commit:
:hammer: Make `BranchDirection` global and simplify `_write_priorities_file()`

- Update tests to avoid usage of "magic numbers" and reference `BranchDirection` where possible

Commit:
:hammer: Remove unnecessary `CPLEXBranchDirection`
